record-core/src/androidMain/kotlin/dev/theolm/record/RecordCore.android.kt

```diff
@@ -4,60 +4,103 @@ package dev.theolm.record
 import dev.theolm.record.error.RecordFailException
+import java.io.FileOutputStream
 
 @Suppress("TooGenericExceptionCaught", "SwallowedException")
 internal actual object RecordCore {
+    private var audioRecord: AudioRecord? = null
+    private var recordingThread: Thread? = null
     private var recorder: MediaRecorder? = null
     private var output: String? = null
-    private var recordingState : RecordingState = RecordingState.IDLE
+    private var myRecordingState : RecordingState = RecordingState.IDLE
 
     @Throws(RecordFailException::class)
     internal actual fun startRecording(config: RecordConfig) {
         checkPermission()
         output = config.getOutput()
-        recorder = createMediaRecorder(config)
+        when(config.outputFormat) {
+            OutputFormat.MPEG_4 -> {
+                recorder = createMediaRecorder(config)
 
-        recorder?.apply {
-            runCatching {
-                prepare()
-            }.onFailure {
-                throw RecordFailException()
-            }
+                recorder?.apply {
+                    runCatching {
+                        prepare()
+                    }.onFailure {
+                        throw RecordFailException()
+                    }
+
+                    setOnErrorListener { mr, what, extra ->
+                        stopRecording(config)
+                    }
 
-            setOnErrorListener { mr, what, extra ->
-                stopRecording()
+                    start()
+                    myRecordingState = RecordingState.RECORDING
+                }
             }
+            OutputFormat.WAV -> {
+                val bufferSize = AudioRecord.getMinBufferSize(
+                    config.sampleRate,
+                    config.outputFormat.toMediaRecorderOutputFormat(),
+                    config.audioEncoder.toMediaRecorderAudioEncoder()
+                )
+
+                audioRecord = AudioRecord(
+                    MediaRecorder.AudioSource.MIC,
+                    config.sampleRate,
+                    config.outputFormat.toMediaRecorderOutputFormat(),
+                    config.audioEncoder.toMediaRecorderAudioEncoder(),
+                    bufferSize
+                )
+
+                audioRecord?.apply {
+                    startRecording()
+                    myRecordingState = RecordingState.RECORDING
 
-            start()
-            recordingState = RecordingState.RECORDING
+                    recordingThread = Thread {
+                        writeAudioDataToFile(bufferSize, config.sampleRate)
+                    }.apply { start() }
+                }
+            }
         }
     }
 
     @Throws(NoOutputFileException::class)
-    internal actual fun stopRecording(): String {
-        recordingState = RecordingState.IDLE
-        recorder?.apply {
-            stop()
-            release()
-        }
+    internal actual fun stopRecording(config: RecordConfig): String {
```

Comment:
`stopRecording` does not need to receive the config via parameter.
instead of changing the change signature use a global var store the config during the startRecording and in the stopRecording you use this config and in the teardown set it to null.

Reply:
sure will do that...